rimage/ext_manifest.c

```diff
@@ -81,16 +82,117 @@ static int ext_man_validate(uint32_t section_size, const void *section_data)
 	}
 }
 
+static int ext_man_read_fw_ready(const struct image *image, uint8_t **dst_buff)
+{
+	const struct module *module;
+	const Elf32_Shdr *section;
+	int read;
+	int i;
+
+	for (i = 0; i < image->num_modules; i++) {
+		module = &image->module[i];
+
+		if (module->fw_ready_index < 0)
+			continue;
+
+		section = &module->section[module->fw_ready_index];
+
+		/* alloc buffer for fw_ready section */
+		*dst_buff = calloc(1, section->size);
+		if (!*dst_buff)
+			return -ENOMEM;
+
+		/* fill buffer with fw_ready section content */
+		fseek(module->fd, section->off, SEEK_SET);
+		read = fread(*dst_buff, 1, section->size, module->fd);
+
+		if (read != section->size) {
+			fprintf(stderr,
+				"error: can't read fw_ready section %d\n",
+				-errno);
+			free(*dst_buff);
+			return -errno;
+		}
+
+		return section->size;
+	}
+
+	return -EINVAL;
+}
+
+static int ext_man_build_window_elem(const struct image *image,
+				     uint8_t **dst_buff)
+{
+	struct ext_man_elem_header ext_man_hdr;
+	struct sof_ipc_ext_data_hdr hdr;
+	uint8_t *buffer = NULL;
+	int ret = -EINVAL;
+	int section_size;
+	int read = 0;
+
+	section_size = ext_man_read_fw_ready(image, &buffer);
+
+	if (section_size < 0) {
+		fprintf(stderr, "warning: section `fw_ready` reading error, code: %d\n",
+			section_size);
+		return section_size;
+	}
+
+	while (read < section_size) {
+		memcpy(&hdr, &buffer[read], sizeof(hdr));
+		if (hdr.type != SOF_IPC_EXT_WINDOW) {
+			read += hdr.hdr.size;
+			continue;
+		}
+
+		ext_man_hdr.type = EXT_MAN_ELEM_WINDOW;
+		ext_man_hdr.elem_size = ALIGN_UP(hdr.hdr.size +
+						 sizeof(ext_man_hdr),
+						 EXT_MAN_ALIGN);
+
+		/* alloc buffer for fw_ready section */
+		*dst_buff = calloc(1, ext_man_hdr.elem_size);
+		if (!(*dst_buff)) {
+			ret = -ENOMEM;
+			break;
+		}
+
+		/* fill buffer with fw_ready section content */
+		memcpy(*dst_buff, &ext_man_hdr, sizeof(ext_man_hdr));
+		memcpy(*dst_buff + sizeof(ext_man_hdr), buffer, hdr.hdr.size);
+
```

Comment:
It would be good to log each section as we add it.

Reply:
@lgirdwood are you talking about something like this?

> 00:14:29,736 INFO  - Extended manifest found module, type: 0x0000 size: 0x0050 ( 80)
> 00:14:29,736 INFO  - Extended manifest found module, type: 0x0001 size: 0x00C0 ( 192)
> 00:14:29,736 INFO  - Extended manifest found module, type: 0x0002 size: 0x0060 ( 96)
> 00:14:29,736 INFO  - Extended manifest found module, type: 0x0003 size: 0x0030 ( 48)
> 00:14:29,736 INFO  - Extended manifest saved to file sof-apl.ri.xman size 432 (0x1B0) bytes

Reply:
ok, great - also best to include offset for anyone who needs to peek at the data.

Reply:
added